gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -166,11 +166,22 @@ def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
 
     current_declaration = T::Private::DeclState.current.active_declaration
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
+    current_mod_final = T::Private::Final.final_module?(mod)
     original_method = mod.instance_method(method_name)
 
+    if !current_declaration.nil? && current_declaration.final && current_mod_final
+      raise "`#{mod.name}` was declared as final and its method `#{method_name}` " \
+        "cannot also be explicitly declared as final (it is implicitly final)"
+    end
     _check_final_ancestors(mod, mod.ancestors, [method_name])
 
-    return if current_declaration.nil?
+    if current_declaration.nil?
+      if current_mod_final
+        add_final_method(method_to_key(original_method))
+        add_module_with_final(mod)
+      end
+      return
+    end
```

Comment:
requiring that it be marked `final` might also simplify this check (though it might also make it more complicated).